downstairs/src/dynamometer.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2023 Oxide Computer Company
+use super::*;
+
+pub enum DynoFlushConfig {
+    FlushPerIops(usize),
+    FlushPerBlocks(usize),
+    FlushPerMs(usize),
+}
+
+pub async fn dynamometer(
+    mut region: Region,
+    num_writes: usize,
+    samples: usize,
+    flush_config: DynoFlushConfig,
+) -> Result<()> {
+    // TODO: pull into another crate? this is copied from measure-iops tool
+    let mut io_operations_sent = 0;
+    let mut bw_consumed = 0;
+    let mut measurement_time = Instant::now();
+    let mut total_io_time = Duration::ZERO;
+    let mut iops: Vec<f32> = vec![];
+    let mut bws: Vec<f32> = vec![];
+
+    let mut flush_number = 0;
+    let mut gen_number = 0;
+    let mut flush_time = Instant::now();
+    let mut iops_since_last_flush = 0;
+    let mut blocks_since_last_flush = 0;
+
+    let ddef = region.def();
+    eprintln!("{:?}", ddef);
+
+    // Fill test: write bytes in whole region
+
+    let mut rng = SmallRng::from_entropy();
+
+    'outer: loop {
+        let block = (0..ddef.block_size() as usize)
+            .map(|_| rng.sample(rand::distributions::Standard))
+            .collect::<Vec<u8>>();
+        let nonce = (0..12)
+            .map(|_| rng.sample(rand::distributions::Standard))
+            .collect::<Vec<u8>>();
+        let tag = (0..16)
+            .map(|_| rng.sample(rand::distributions::Standard))
+            .collect::<Vec<u8>>();
+
+        let hash = integrity_hash(&[&nonce, &tag, &block]);
+
+        for eid in 0..ddef.extent_count() {
+            let mut block_offset = 0;
+            loop {
+                if block_offset >= ddef.extent_size().value {
+                    break;
+                }
+
+                let block = block.clone();
+                let nonce = nonce.clone();
+                let tag = tag.clone();
+
+                let writes: Vec<_> = (0..num_writes)
+                    .map(|i| crucible_protocol::Write {
+                        eid: eid as u64,
+                        offset: Block::new_with_ddef(
+                            i as u64 + block_offset,
+                            &ddef,
+                        ),
+                        data: bytes::Bytes::from(block.clone()),
```

Comment:
This adds an extra clone of the full data; I think it would be faster if we constructed the `Bytes` object outside the loop (since it's reference-counted and cheaper to clone here).

Reply:
:+1: moved out in 46d7d49